### DISCUSSION TITLE
Pattern matching: override deconstruct

### DISCUSSION THREAD

**Discussion by akiraKido** [dotnet/csharplang#2852] [Language Ideas]
For pattern matching, I'd like to deconstruct a certain struct I made as follows:

	public struct Result<T>
	{
		public readonly bool IsSuccess;
		public readonly T Value;
		public readonly string? Message;

		private Result(bool success, T value, string? message) =>
			(IsSuccess, Value, Message) = (success, value, message);

		public void Deconstruct(out bool success, out T value, out string? message) =>
			(success, value, message) = (IsSuccess, Value, Message);

		public void Deconstruct(out bool success, out string? message) =>
			(success, message) = (IsSuccess, Message);
		public void Deconstruct(out bool success, out T value) =>
			(success, value) = (IsSuccess, Value);

		public static Result<T> Fail(string message) => new Result<T>(false, default, message);
		public static Result<T> Success(T value) => new Result<T>(true, value, null);
	}

so that I would be able to do something like this:

	switch (TryGetValue(ref input, ref length))
	{
		case (false, string e):
			return Result<JsonNode>.Fail(e);
		case (true, JsonNode value):
			values.Add(value);
			break;
	}

This currently doesn't compile, saying that the deconstruction is ambiguous, despite explicitly declaring their types.

This works:

	switch (TryGetValue(ref input, ref length))
	{
		case (false, _, var e):
			return Result<JsonNode>.Fail(e);
		case (true, var value, _):
			values.Add(value);
			break;
	}

but the `_` seems verbose to me.

**Comment by HaloFour** [comment]
Deconstruct can only be overloaded by arity of the `out` parameters.  The compiler needs to be able to resolve the deconstructor before it can apply any of the positional patterns, including type parameters.
